sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnusedPrivateMember.cs

```diff
@@ -80,462 +65,182 @@ protected sealed override void Initialize(SonarAnalysisContext context)
+                                removableSymbolsCollector.FieldLikeSymbols);
+                            foreach (var diagnostic in diagnostics)
+                            {
+                                cc.ReportDiagnosticIfNonGenerated(diagnostic, cc.Compilation);
+                            }
                         },
                         SymbolKind.NamedType);
 
                     c.RegisterCompilationEndAction(
-                       cc =>
-                       {
-                           if (!shouldRaise)
-                           {
-                               return;
-                           }
-
-                           foreach (var diagnostic in possibleUnusedInternalMembers)
-                           {
-                               cc.ReportDiagnosticIfNonGenerated(diagnostic, cc.Compilation);
-                           }
-                       });
+                        cc =>
+                        {
+                            var foundInternalsVisibleTo = cc.Compilation.Assembly
+                                .GetAttributes(KnownType.System_Runtime_CompilerServices_InternalsVisibleToAttribute)
+                                .Any();
+
+                            if (foundInternalsVisibleTo)
+                            {
+                                return;
+                            }
+
+                            var usageCollector = new SymbolUsageCollector(
+                                c.Compilation.GetSemanticModel,
+                                removableInternalMembers.Select(s => s.Name).ToHashSet());
+
+                            foreach (var symbol in allNamedTypes)
+                            {
+                                VisitDeclaringReferences(symbol, usageCollector);
+                            }
+
+                            var diagnostics = GetDiagnostics(usageCollector, removableInternalMembers.ToHashSet(), "internal",
+                                new BidirectionalDictionary<ISymbol, SyntaxNode>());
+                            foreach (var diagnostic in diagnostics)
+                            {
+                                cc.ReportDiagnosticIfNonGenerated(diagnostic, cc.Compilation);
+                            }
+                        });
                 });
         }
 
-        private static bool IsInternalVisibleToAttribute(AttributeSyntax attribute, SemanticModel semanticModel)
+        private static IEnumerable<Diagnostic> GetDiagnostics(SymbolUsageCollector usageCollector, ISet<ISymbol> removableSymbols,
+            string accessibility, BidirectionalDictionary<ISymbol, SyntaxNode> fieldLikeSymbols)
         {
+            var unusedSymbols = removableSymbols
+                .Except(usageCollector.UsedSymbols)
+                .Where(symbol => !MentionedInDebuggerDisplay(symbol))
+                .ToHashSet();
+
+            var propertiesWithUnusedAccessor = removableSymbols

```

Comment:
Why is this value calculated here and passed in to _GetDiagnosticsForPropertyAccessors_, rather than just being calculated in _GetDiagnosticsForPropertyAccessors_?

Reply:
The code is not much better when the collection is moved inside the other method

Reply:
Also, when you mentioned the `MentionedInDebuggerDisplay`, I remember it is actually the reason for calculating the collections outside...